operator_manifest/cli.py

```diff
@@ -0,0 +1,285 @@
+
+    :param file replacements_file: the file-like object to the replacements file. The format of
+        this file is a simple JSON object where each attribute is a string representing the
+        original image reference and the value is a string representing the new value for the
+        image reference
+    :param str manifest_dir: the path to the directory where the manifest files are stored
+    :param bool dry_run: whether or not to apply the replacements
+    :raises ValueError: if more than on CSV in manifest_dir
+    :raises ValueError: if validation fails
+    """
+    abs_manifest_dir = _normalize_dir_path(manifest_dir)
+    logger.info(
+        'Replacing image references in CSVs found in %s with replacements from %s',
+        abs_manifest_dir,
+        replacements_file.name,
+    )
+
+    operator_manifest = OperatorManifest.from_directory(abs_manifest_dir)
+
+    if not _should_apply_replacements(manifest_dir):
+        logger.warning('Skipping replacements')
+        return
+
+    replacements = {}
+    for k, v in json.load(replacements_file).items():
+        replacements[ImageName.parse(k)] = ImageName.parse(v)
+        logger.info('%s -> %s', k, v)
+
+    operator_manifest.csv.replace_pullspecs_everywhere(replacements)
+
+    logger.info('Setting related images section')
+    operator_manifest.csv.set_related_images()
+
+    if not dry_run:
+        operator_manifest.csv.dump()
+        logger.info('Image references replaced')
+
+
+def pin_image_references(
+    manifest_dir,
+    authfile=None,
+    output_extract=None,
+    output_replace=None,
+    dry_run=False,
+):
+    """
+    Pins to digest all the image references from the CSVs found in manifest_dir.
+
+    For each image reference, if a tag is used, it is resolved to a digest by querying the
+    container image registry. Then, each reference is replaced with the resolved, pinned, version.
+
+    :param str manifest_dir: the path to the directory where the manifest files are stored
+    :param str authfile: the path to the authentication file for registry communication
+    :param file output_extract: the file-like object to store the extracted image references
+    :param file output_replace: the file-like object to store the image reference replacements
+    :param bool dry_run: whether or not to apply the replacements
+    :raises ValueError: if more than on CSV in manifest_dir
+    :raises ValueError: if validation fails
+    """
+    if not output_replace.seekable():
+        raise ValueError('output_replace must be a seekable object')
+
+    references = extract_image_references(manifest_dir, output=output_extract)
+
+    if not _should_apply_replacements(manifest_dir):
+        logger.warning('Skipping replacements. Replacement file is not created')
+        return
+
+    replacements = {}
+    for reference in references:
+        replacements[reference] = resolve_bundle_digest(reference, authfile=authfile)
```

Comment:
This is wrong. When a bundle image is resolved, we should be done to an Image Manifest. This is because bundle images are non-binary and arch-independent. However! The images we are resolving here are not bundle images. They are operator and operand images, aka related images. These are very much binary. I'll fix this.

Reply:
Addressed this.